docs/proposals/ingesters-migration.md

```diff
@@ -0,0 +1,60 @@
+* Shutdown single chunks-ingester, run flusher in its place, and when done start new blocks ingester. This is similar to the procedure we did during the introduction of WAL. Flusher can be run via initContainer support in pods. This still requires two-step deployment: 1) enable flusher and reconfigure ingesters to use blocks, 2) remove flusher.
+
+After all ingesters are converted to blocks, we can set cut-off time for querying chunks storage on queriers.
+
+For rollback from blocks to chunks, we need to be able to flush data from ingesters to the blocks storage, and then switch ingesters back to chunks. Ingesters are currently not able to flush blocks to storage, but adding flush-on-shutdown option, support for `/shutdown` endpoint and support in flusher component similar to chunks is doable, and should be part of this work.
+
+With this ability, rollback would follow the same process, just in reverse: 1) redeploy with flush flag enabled, 2) redeploy with config change from blocks to chunks.
+
+Alternative plan could be to use a separate Cortex cluster configured to use blocks, and redirect incoming traffic to both chunks and blocks cluster. When one is confident about the blocks cluster running correctly, old chunks cluster can be shutdown. In this plan, there is an overlap where both clusters are ingesting same data. Blocks cluster needs to be configured to be able to query chunks storage as well, with cut-off time based on when clusters were configured (at latest, to minimize amount of duplicated samples that need to be processed during queries.)
+
+## Querying
+
+To be able to query both blocks and chunks, querier needs to be extended with this ability, and merge results from both. Queriers can always query blocks – each querier knows about existing blocks and their timeranges, so it can quickly determine whether there are any blocks with relevant data. Always querying blocks is also useful when there is some background process converting chunks to blocks. As new blocks appear on the store, they get queried even if they are old.
+
+For querying chunks storage, we have two options:
+- Always query the chunks store – useful during ingesters switch, or after rollback from blocks to chunks.
+- Query chunk store only for queries that ask for data after specific cut-off time. This is useful after all ingesters have switched, and we know the timestamp since ingesters are only writing blocks.
+
+Querier needs to support both. Switching between them via config option seems fine. (We could use runtime-config for on-the-fly switch without restarts)
```

Comment:
Queriers are stateless and restarting them is quick. I don't think we need to use the runtime-config for that.